How many late days are you using for this assignment? 

In [1]:
# Imports and pip installations (if needed)
import pandas as pd


# Part 1: Load the dataset

In [3]:
# Load the given datasets
dataSet1 = pd.read_csv("/Users/<redacted>/Downloads/chronic_kidney_disease_categorical.csv")
dataSet2 = pd.read_csv("/Users/<redacted>/Downloads/chronic_kidney_disease_numerical.csv")

# Print the data
print(dataSet1)
print(dataSet2)


     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The Chronic Kidney Disease (CKD) dataset from the UCI repository contains records that capture various health measurements for patients. The dataset includes features such as patient age, blood pressure, and lab test results along with qualitative assessments like red blood cell quality and pus cell presence. The label in the dataset indicates whether a patient has chronic kidney disease, making it a binary classification problem. Features in the dataset are of both types: numerical and categorical. Numerical examples include measurements like age, blood pressure, and blood glucose levels, which provide continuous data. Categorical examples involve indicators like red blood cell quality, pus cell quality, and the presence of hypertension, typically marked as 'normal' or 'abnormal' or yes/no. Each patient record is a composite of these varied health parameters, allowing for analysis of how each factor may contribute to kidney disease. Overall, the dataset is well-suited for machine learning applications that require careful preprocessing to handle the mix of numerical and categorical data.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [71]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates1 = dataSet1.duplicated().sum()
if(duplicates1 > 0):
    print(duplicates1)


# Drop these duplicate rows
dataSet1_no_dups = dataSet1.drop_duplicates()


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
# duplicates2 = dataSet2.duplicated()
# if(duplicates1 > 0):
#     print(duplicates1)


duplicates2 = dataSet2.duplicated().sum()
if(duplicates2 > 0):
    print(dataSet2.duplicated())
    
dataSet2_no_dups = dataSet2.drop_duplicates()





30
0      False
1      False
2      False
3      False
4      False
       ...  
415    False
416    False
417    False
418    False
419    False
Length: 420, dtype: bool


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

In [72]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
from IPython.display import display
merged = pd.merge(dataSet1_no_dups, dataSet2_no_dups, on= "unique_id")
#Print the combined dataset
print(merged)
#display(merged)



     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
385     789906  0.0  0.0    normal    normal  notpresent  notpresent   no   
386     490253  0.0  0.0    normal    normal  notpresent  notpresent   no   
387     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes   
388     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   
389     583971  1.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm cad appet   pe  ane  Target   age    bp    bgr     bu   sc    sod 

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [95]:
# Calculate the percentage of rows that contain atleast one missing value
total = merged.shape[0]
missing = merged[merged.isnull().any(axis=1)].shape[0]
percent = missing/total
# Print %
print(percent)

# Drop these rows from the dataset
merged_cleaned = merged.dropna()

# Print the Dataset
print(merged_cleaned)

0.6076923076923076
     unique_id   al   su       rbc      pc         pcc          ba  htn   dm  \
4       881763  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
5       114717  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
7       901528  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
8       433613  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
9       621332  1.0  0.0  abnormal  normal  notpresent  notpresent   no  yes   
..         ...  ...  ...       ...     ...         ...         ...  ...  ...   
382     996076  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
384     397388  2.0  2.0    normal  normal  notpresent     present  yes   no   
385     789906  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
386     490253  0.0  0.0    normal  normal  notpresent  notpresent   no   no   
388     273038  0.0  0.0    normal  normal  notpresent  notpresent   no   no   

     cad appet  pe a

## Part 3.4: Sort the dataset according to the Labels

In [74]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = (merged_cleaned.sort_values(by="Target")).reset_index(drop="true")


# Print the dataset
print(sorted_df)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       534245  2.0  0.0    normal  abnormal     present  notpresent   no   
1       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
4       514721  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     489160  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     530138  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     261137  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     322882  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     273038  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad appet   pe  ane  Target   age    bp    bgr     bu    sc    so

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [75]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_df = pd.get_dummies(sorted_df, columns=categorical_columns, dtype=int)

# Print the dataset
print(encoded_df)

     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  hemo  \
0       534245  2.0  0.0  45.0  70.0  117.0   52.0   2.2  136.0  3.8  10.0   
1       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7   
3       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1   
4       514721  2.0  0.0  56.0  90.0  129.0  107.0   6.7  131.0  4.8   9.1   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...   ...   
148     489160  0.0  0.0  28.0  60.0   79.0   50.0   0.5  145.0  5.0  17.6   
149     530138  0.0  0.0  23.0  80.0  111.0   34.0   1.1  145.0  4.0  14.3   
150     261137  0.0  0.0  35.0  60.0  105.0   39.0   0.5  135.0  3.9  14.7   
151     322882  0.0  0.0  33.0  80.0  128.0   38.0   0.6  135.0  3.9  13.1   
152     273038  0.0  0.0  47.0  60.0  117.0   22.0   1.2  138.0  3.5  13.0   

      pcv     wbcc  rbcc  rbc_abnormal  rbc_normal  pc_abnormal

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Using a single binary column for a categorical variable with more than two values would not be equivalent in downstream modeling. Separate binary columns for each category is better because it prevents the model from assuming an incorrect ordinal relationship between categories. This approach ensures each category is treated independently, leading to more accurate predictions.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [96]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
df_copy = encoded_df.copy()
for col in numerical_columns:
    standard1 = encoded_df[col].std()
    mean1 = encoded_df[col].mean()
    lower = mean1 - 3*standard1
    upper = mean1 + 3*standard1
    conditional = (encoded_df[col] >= lower) & (encoded_df[col] <= upper)
    df_copy = df_copy[conditional]


cleaned = df_copy.reset_index(drop=True)

# Print the dataset
print(cleaned)

     unique_id   al   su   age    bp    bgr     bu   sc    sod  pot  hemo  \
0       343710  2.0  0.0  61.0  80.0  173.0  148.0  3.9  135.0  5.2   7.7   
1       514721  2.0  0.0  56.0  90.0  129.0  107.0  6.7  131.0  4.8   9.1   
2       124304  3.0  4.0  69.0  70.0  264.0   87.0  2.7  130.0  4.0  12.5   
3       980291  3.0  0.0  62.0  70.0  122.0   42.0  1.7  136.0  4.7  12.6   
4       615697  4.0  0.0   6.0  60.0   94.0   67.0  1.0  135.0  4.9   9.9   
..         ...  ...  ...   ...   ...    ...    ...  ...    ...  ...   ...   
129     489160  0.0  0.0  28.0  60.0   79.0   50.0  0.5  145.0  5.0  17.6   
130     530138  0.0  0.0  23.0  80.0  111.0   34.0  1.1  145.0  4.0  14.3   
131     261137  0.0  0.0  35.0  60.0  105.0   39.0  0.5  135.0  3.9  14.7   
132     322882  0.0  0.0  33.0  80.0  128.0   38.0  0.6  135.0  3.9  13.1   
133     273038  0.0  0.0  47.0  60.0  117.0   22.0  1.2  138.0  3.5  13.0   

      pcv     wbcc  rbcc  rbc_abnormal  rbc_normal  pc_abnormal  pc_normal 

/var/folders/fs/_38gxvr15n988sw8z7f0lfj00000gn/T/ipykernel_42403/1769173375.py:10: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_copy = df_copy[conditional]
/var/folders/fs/_38gxvr15n988sw8z7f0lfj00000gn/T/ipykernel_42403/1769173375.py:10: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_copy = df_copy[conditional]
/var/folders/fs/_38gxvr15n988sw8z7f0lfj00000gn/T/ipykernel_42403/1769173375.py:10: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_copy = df_copy[conditional]
/var/folders/fs/_38gxvr15n988sw8z7f0lfj00000gn/T/ipykernel_42403/1769173375.py:10: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_copy = df_copy[conditional]
/var/folders/fs/_38gxvr15n988sw8z7f0lfj00000gn/T/ipykernel_42403/1769173375.py:10: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df_copy = df_copy[conditional]
/var/folders/fs/_38gxvr15n988sw8z7f0lfj0

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [89]:
# Normalize the all Numerical Attributes in the dataset.

df_new_copy = df_copy.copy()
for col in numerical_columns:
    df_new_copy[col] = (df_copy[col] - df_copy[col].min() / df_copy[col].max() - df_copy[col].min())
# Print the dataset
df_new_copy = df_new_copy.reset_index(drop="true")
print(df_new_copy)

     unique_id   al   su     age    bp         bgr         bu        sc   sod  \
0       343710  2.0  0.0  54.925  19.4  102.768977  137.93865  3.452941  14.2   
1       514721  2.0  0.0  49.925  29.4   58.768977   96.93865  6.252941  10.2   
2       124304  3.0  4.0  62.925   9.4  193.768977   76.93865  2.252941   9.2   
3       980291  3.0  0.0  55.925   9.4   51.768977   31.93865  1.252941  15.2   
4       615697  4.0  0.0  -0.075  -0.6   23.768977   56.93865  0.552941  14.2   
..         ...  ...  ...     ...   ...         ...        ...       ...   ...   
129     489160  0.0  0.0  21.925  -0.6    8.768977   39.93865  0.052941  24.2   
130     530138  0.0  0.0  16.925  19.4   40.768977   23.93865  0.652941  24.2   
131     261137  0.0  0.0  28.925  -0.6   34.768977   28.93865  0.052941  14.2   
132     322882  0.0  0.0  26.925  19.4   57.768977   27.93865  0.152941  14.2   
133     273038  0.0  0.0  40.925  -0.6   46.768977   11.93865  0.752941  17.2   

          pot       hemo   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: I excluded the column about unique_id because a patient's unique_id would never have an impact on if they have ckd or not, it is irrelevant to the disease. The rest of the columns are include health, diet, personal information that may have a correlation to ckd's presence or not, thus i kept them.

In [87]:
#Remove that column
cleaned_again = df_new_copy.drop(columns='unique_id')


# Print the dataset
print(cleaned_again)

      al   su     age    bp         bgr         bu        sc   sod       pot  \
0    2.0  0.0  54.925  19.4  102.768977  137.93865  3.452941  14.2  1.846875   
1    2.0  0.0  49.925  29.4   58.768977   96.93865  6.252941  10.2  1.446875   
2    3.0  4.0  62.925   9.4  193.768977   76.93865  2.252941   9.2  0.646875   
3    3.0  0.0  55.925   9.4   51.768977   31.93865  1.252941  15.2  1.346875   
4    4.0  0.0  -0.075  -0.6   23.768977   56.93865  0.552941  14.2  1.546875   
..   ...  ...     ...   ...         ...        ...       ...   ...       ...   
129  0.0  0.0  21.925  -0.6    8.768977   39.93865  0.052941  24.2  1.646875   
130  0.0  0.0  16.925  19.4   40.768977   23.93865  0.652941  24.2  0.646875   
131  0.0  0.0  28.925  -0.6   34.768977   28.93865  0.052941  14.2  0.546875   
132  0.0  0.0  26.925  19.4   57.768977   27.93865  0.152941  14.2  0.546875   
133  0.0  0.0  40.925  -0.6   46.768977   11.93865  0.752941  17.2  0.146875   

          hemo        pcv          wbcc

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [81]:
# Export the dataframe to a new csv file

csv_file_path = '/Users/<redacted>/Downloads/cleaned_data.csv' 
cleaned_again.to_csv(csv_file_path, index=False)

# Export the DataFrame to a new JSON file
json_file_path = '/Users/<redacted>/Downloads/cleaned_data.json'  
cleaned_again.to_json(json_file_path, orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 
|    al |   su |      age |    bp |         bgr |         bu |        sc |   sod |      pot |     hemo |        pcv |       wbcc |   rbcc |   rbc_abnormal |   rbc_normal |
|-------|------|----------|-------|-------------|------------|-----------|-------|----------|-----------|------------|------------|--------|----------------|--------------|
|   2.0 |  0.0 |  54.925  |  19.4 |  102.768977 |  137.93865 |  3.452941 |  14.2 |  1.846875 |  1.785393 |   7.703704 |  4899.742515 |    0.2 |              1 |            0 |
|   2.0 |  0.0 |  49.925  |  29.4 |   58.768977 |   96.93865 |  6.252941 |  10.2 |  1.446875 |  3.185393 |  12.703704 |  2099.742515 |    0.4 |              1 |            0 |
|   3.0 |  4.0 |  62.925  |   9.4 |  193.768977 |   76.93865 |  2.252941 |   9.2 |  0.646875 |  6.585393 |  20.703704 |  5299.742515 |    1.1 |              0 |            1 |
|   3.0 |  0.0 |  55.925  |   9.4 |   51.768977 |   31.93865 |  1.252941 |  15.2 |  1.346875 |  6.685393 |  22.703704 |  3599.742515 |    0.9 |              0 |            1 |
|   4.0 |  0.0 |  -0.075  |  -0.6 |   23.768977 |   56.93865 |  0.552941 |  14.2 |  1.546875 |  3.985393 |  13.703704 | 12399.742515 |    1.8 |              1 |            0 |
|   3.0 |  1.0 |  48.925  |  19.4 |  143.768977 |   62.93865 |  3.452941 |  16.2 |  1.546875 |  4.985393 |  17.703704 |  3099.742515 |    0.7 |              0 |            1 |
|   1.0 |  0.0 |  39.925  |  -0.6 |   92.768977 |   81.93865 |  2.852941 |  20.2 |  0.646875 |  3.885393 |  11.703704 | 10299.742515 |    0.2 |              0 |            1 |
|   4.0 |  2.0 |  57.925  |  39.4 |   92.768977 |   43.93865 |  6.752941 |  19.2 |  1.246875 |  5.185393 |  20.703704 |  4899.742515 |    1.1 |              0 |            0 |
|   4.0 |  0.0 |  41.925  |  19.4 |   24.768977 |  152.93865 |  7.252941 |  15.2 |  0.446875 |  6.585393 |  12.703704 |  1599.742515 |    0.9 |              1 |            1 |
|   4.0 |  3.0 |  62.925  |   9.4 |  143.768977 |   85.93865 |  5.852941 |  -0.8 |  0.546875 |  3.285393 |   8.703704 |  2689.742515 |    0.4 |              1 |            0 |
|   3.0 |  2.0 |  66.925  |  39.4 |  224.768977 |   79.93865 |  5.152941 |  19.2 | -0.453125 |  4.185393 |  15.703704 |  2899.742515 |    0.6 |              0 |            1 |
|   4.0 |  0.0 |  14.925  |  29.4 |   36.768977 |   29.93865 |  1.252941 |   4.2 |  0.146875 |  2.485393 |   5.703704 |  1989.742515 |    0.3 |              0 |            1 |
|   4.0 |  1.0 |  57.925  |  -0.6 |  168.768977 |   47.93865 |  3.852941 |  16.2 |  2.046875 |  2.385393 |   5.703704 |  3489.742515 |    0.4 |              1 |            0 |
|   4.0 |  1.0 |  49.925  |   9.4 |  139.768977 |   15.93865 |  1.252941 |  15.2 |  0.446875 |  2.685393 |   9.703704 | 2699.742515 |    0.5 |              0 |            1 |
|   4.0 |  0.0 |  41.925  |  19.4 |   62.768977 |  128.93865 |  8.052941 |  11.2 |  2.146875 |  2.885393 |   5.703704 |  1999.742515 |    0.1 |              1 |            0 |

### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [83]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns using the mapping
cleaned_again.rename(columns=rename_mapping, inplace=True)


     albumin  sugar     age  blood pressure  blood glucose random  blood urea  \
0        2.0    0.0  54.925            19.4            102.768977   137.93865   
1        2.0    0.0  49.925            29.4             58.768977    96.93865   
2        3.0    4.0  62.925             9.4            193.768977    76.93865   
3        3.0    0.0  55.925             9.4             51.768977    31.93865   
4        4.0    0.0  -0.075            -0.6             23.768977    56.93865   
..       ...    ...     ...             ...                   ...         ...   
129      0.0    0.0  21.925            -0.6              8.768977    39.93865   
130      0.0    0.0  16.925            19.4             40.768977    23.93865   
131      0.0    0.0  28.925            -0.6             34.768977    28.93865   
132      0.0    0.0  26.925            19.4             57.768977    27.93865   
133      0.0    0.0  40.925            -0.6             46.768977    11.93865   

     serum creatinine  sodi

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [90]:
# Converted SQL to Pandas code
result = cleaned_again.groupby('Target_ckd').size().reset_index(name='count')
print(result)

   Target_ckd  count
0           0    111
1           1     23


*Describe what the above code does here*
The code basically counts for how many patients have ckd or not. The 0 means that a patient doesnt have ckd and the count next to it displays the number of patients without ckd. So 111 patients without ckd. The 1 in target_ckd column means that a patient does have ckd and the count for that is 23 patients.